[morphologica] new port

rremilian opened the pull request:
Fixes #43719

- [X] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [X] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [X] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [X] The versioning scheme in `vcpkg.json` matches what upstream says.
- [X] The license declaration in `vcpkg.json` matches what upstream says.
- [X] The installed as the "copyright" file matches what upstream says.
- [X] The source code of the component installed comes from an authoritative source.
- [X] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [X] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [X] Only one version is in the new port's versions file.
- [X] Only one version is added to each modified port's versions file.

rremilian committed:
[morphologica] new port

rremilian commented:
How can I check the detailed build logs produced by the pipelines ?
I cannot reproduce the error on `x64-linux`.

rremilian committed:
[morphologica] add fix_nlohmann_include.patch

rremilian committed:
[morphologica] Windows builds have not been tested since 2021

rremilian commented:
@Mengna-Li I think the only issue on Linux is that `libgbm.h` is missing. On my local machine, I fixed that by installing `libgbm-devel` (`libgbm-dev`). 
I saw that `libgbm` is part of `mesa`. I tried to add `mesa` to the manifest file, but I find that is a little bit overkill to install `mesa` with all the default features (including `llvm`) just for that (the `default-features: false` will not stop this).
Couldn't we insert a warning message in the portfile which will say that you need to install `libgbm` for this port ? Similar to this:

https://github.com/microsoft/vcpkg/blob/5c41ff21f6feda702a7d5a1afc765f88d25783c5/ports/coin/portfile.cmake#L2-L13

rremilian committed:
[morphologica] add freeglut

rremilian committed:
[morphologica] add prevent examples patch

rremilian committed:
[morphologica] remove egl_fix.patch (only needed for Windows)

rremilian committed:
[morphologica] prevent tests and examples

Also, try one more time with Windows.

rremilian committed:
[morpholica] add back egl_fix.patch for Windows

rremilian committed:
[morphologica] remove support for Windows

rremilian committed:
[morphologica] use freeglut only on Linux

rremilian commented:
> @Mengna-Li I think the only issue on Linux is that `libgbm.h` is missing. On my local machine, I fixed that by installing `libgbm-devel` (`libgbm-dev`). I saw that `libgbm` is part of `mesa`. I tried to add `mesa` to the manifest file, but I find that is a little bit overkill to install `mesa` with all the default features (including `llvm`) just for that (the `default-features: false` will not stop this). Couldn't we insert a warning message in the portfile which will say that you need to install `libgbm` for this port ? Similar to this:
> 
> https://github.com/microsoft/vcpkg/blob/5c41ff21f6feda702a7d5a1afc765f88d25783c5/ports/coin/portfile.cmake#L2-L13
> 
> Later Edit: And also these: `libglu1-mesa-dev libxmu-dev libxi-dev`
> 
> Later Edit 2: I see that `libgbm` is needed only for one example program. I think I can work around that.

I think I have solved this. I will check it tomorrow.
On `osx` everything seems good.
From what I saw on upstream, Windows is not actively supported. Disabling tests and examples helped to go pass some errors related to C++20 standard. However, some functions are not defined on Windows (#ifndef __WIN__) and also doesn't seem to have an equivalent yet.

Mengna-Li commented on ports/morphologica/portfile.cmake:
```diff
@@ -0,0 +1,21 @@
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/lib")
+
+vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE.txt")
+
```
There is an extra blank line here.

rremilian replied:
Thanks for the review. Fixed.

Mengna-Li commented on ports/morphologica/portfile.cmake:
```diff
@@ -0,0 +1,21 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO ABRG-Models/morphologica
+    REF v${VERSION}
```
```suggestion
    REF "v${VERSION}"
```

Mengna-Li commented on ports/morphologica/remove_number_type.patch:
```diff
@@ -0,0 +1,13 @@
+ # Header installation
+ install(
+-  FILES quaternion.h tools.h BezCoord.h BezCurve.h BezCurvePath.h ReadCurves.h AllocAndRead.h MorphDbg.h mathconst.h MathAlgo.h MathImpl.h geometry.h number_type.h Hex.h HexGrid.h hexyhisto.h CartGrid.h histo.h keys.h Grid.h HdfData.h Process.h RD_Base.h DirichVtx.h DirichDom.h ShapeAnalysis.h NM_Simplex.h Rect.h Anneal.h Config.h vec.h vvec.h mat22.h mat33.h mat44.h colour.h ColourMap.h ColourM ...
++  FILES quaternion.h tools.h BezCoord.h BezCurve.h BezCurvePath.h ReadCurves.h AllocAndRead.h MorphDbg.h mathconst.h MathAlgo.h MathImpl.h geometry.h Hex.h HexGrid.h hexyhisto.h CartGrid.h histo.h keys.h Grid.h HdfData.h Process.h RD_Base.h DirichVtx.h DirichDom.h ShapeAnalysis.h NM_Simplex.h Rect.h Anneal.h Config.h vec.h vvec.h mat22.h mat33.h mat44.h colour.h ColourMap.h ColourMap_Lists.h len ...
```
What issues did this patch fix? Can you submit this change to the upstream?

rremilian replied:
The file no longer exists.
In this [commit](https://github.com/ABRG-Models/morphologica/commit/4051896f47120dc62f6dbc778022324f141b116f), `number_type.h` was moved into `traits_tests.h`, but it's still mentioned in `morph/CMakeLists.txt`. I will open an issue to the upstream.

rremilian replied:
This change was applied to the upstream.

rremilian committed:
[morphologica] minor changes

rremilian committed:
[morphologica] fonts are required

rremilian commented:
A new release is on the way. We should wait for it, since we will no longer need the patches.

rremilian committed:
[morphologica] update to v3.3

The patches are no longer needed.
Everything was added in the upstream repository

rremilian committed:
[morphologica] update to 3.3.1

rremilian committed:
[morphologica] add fix_egl.patch

rremilian committed:
[morphologica] use patch from upstream

BillyONeal commented on ports/morphologica/portfile.cmake:
```diff
@@ -0,0 +1,28 @@
+)
+vcpkg_cmake_install()
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug")
```
https://github.com/ABRG-Models/morphologica?tab=readme-ov-file#what-is-morphologica claims 'header-only', could this be replaced with adding `set(VCPKG_BUILD_TYPE release)` to the top of the file?

sebjameswml replied:
You're right that these headers should not be bundled in morphologica. I've now removed all bundled GL headers such as these:
```
morphologica:x64-windows:/include/morph/GL3/gl3.h
morphologica:x64-windows:/include/morph/GL3/gl3w.h
```
These are all a legitimate part of morphologica and *not* third party code: 
```
morphologica:x64-windows:/include/morph/gl/compute_manager.h
morphologica:x64-windows:/include/morph/gl/compute_manager_cli.h
morphologica:x64-windows:/include/morph/gl/compute_shaderprog.h
morphologica:x64-windows:/include/morph/gl/shaders.h
morphologica:x64-windows:/include/morph/gl/ssbo.h
morphologica:x64-windows:/include/morph/gl/texture.h
morphologica:x64-windows:/include/morph/gl/util.h
morphologica:x64-windows:/include/morph/gl/version.h
```
Yes, rapidxml needs to be handled differently and ideally not bundled. I've fixed these upstream:
```
morphologica:x64-windows:/include/morph/rapidxml.hpp
morphologica:x64-windows:/include/morph/rapidxml_iterators.hpp
morphologica:x64-windows:/include/morph/rapidxml_print.hpp
morphologica:x64-windows:/include/morph/rapidxml_utils.hpp
```

rremilian replied:
@BillyONeal I have bumped the version to 3.4 and included `rapidxml` and `egl`.
I've also changed the manifest to build only the Release version.

BillyONeal reviewed (changes requested):
It looks like this port has several vendored dependencies. For example, I think https://github.com/ABRG-Models/morphologica/blob/main/include/EGL/egl.h needs to use https://github.com/microsoft/vcpkg/blob/master/ports/egl/vcpkg.json ?

A few selected installs that look problematic:

```
morphologica:x64-windows:/include/morph/GL3/gl3.h
morphologica:x64-windows:/include/morph/GL3/gl3w.h
morphologica:x64-windows:/include/morph/gl/compute_manager.h
morphologica:x64-windows:/include/morph/gl/compute_manager_cli.h
morphologica:x64-windows:/include/morph/gl/compute_shaderprog.h
morphologica:x64-windows:/include/morph/gl/shaders.h
morphologica:x64-windows:/include/morph/gl/ssbo.h
morphologica:x64-windows:/include/morph/gl/texture.h
morphologica:x64-windows:/include/morph/gl/util.h
morphologica:x64-windows:/include/morph/gl/version.h
morphologica:x64-windows:/include/morph/rapidxml.hpp
morphologica:x64-windows:/include/morph/rapidxml_iterators.hpp
morphologica:x64-windows:/include/morph/rapidxml_print.hpp
morphologica:x64-windows:/include/morph/rapidxml_utils.hpp
```

rremilian committed:
[morphologica] bump to version 3.4

rremilian committed:
[morphologica] update to 3.4.1

rremilian committed:
[morphologica] update to v3.4.3

Mengna-Li commented on versions/m-/morphologica.json:
```diff
@@ -0,0 +1,24 @@
+      "version": "3.4.1",
+      "port-version": 0
+    },
+    {
```
Please ensure that only one version is added.

rremilian replied:
Done. Thanks !

BillyONeal commented:
Thanks for the new port!
